ledger/streaming/client/src/backfill.rs

```diff
@@ -0,0 +1,178 @@
+        self.upstream
+            .get_block_stream(starting_height)
+            .map(|upstream| {
+                backfill_stream(
+                    upstream,
+                    starting_height,
+                    self.fetcher.clone(),
+                    self.logger.clone(),
+                )
+            })
+    }
+}
+
+fn backfill_stream<
+    S: Stream<Item = Result<BlockStreamComponents>> + 'static,
+    F: BlockFetcher + 'static,
+>(
+    upstream: S,
+    starting_height: u64,
+    fetcher: Arc<F>,
+    logger: Logger,
+) -> impl Stream<Item = Result<BlockStreamComponents>> {
+    use futures::stream::{empty, once};
+
+    let mut prev_index: Option<BlockIndex> = None;
+    upstream.flat_map(
+        move |result| -> Pin<Box<dyn Stream<Item = Result<BlockStreamComponents>>>> {
+            let next_index = prev_index.map_or_else(|| starting_height, |index| index + 1);
+            match result {
+                Ok(components) => {
+                    let index = components.block_data.block().index;
+                    if prev_index.is_some() && index <= prev_index.unwrap() {
+                        log::info!(
+                            logger,
+                            "Ignoring streamed block with index={} since last fetched index={}",
+                            index,
+                            prev_index.unwrap()
+                        );
+                        return Box::pin(empty());
+                    }
+
+                    let item_stream = once(async { Ok(components) });
+                    if index == next_index {
+                        prev_index = Some(index);
+                        Box::pin(item_stream)
+                    } else {
+                        let start = prev_index.unwrap_or(starting_height);
+                        prev_index = Some(index);
+                        match fetcher.fetch_range(start..index) {
+                            Ok(backfill) => Box::pin(backfill.chain(item_stream)),
+                            Err(fetch_err) => {
+                                log::warn!(
+                                    logger,
+                                    "Failed to backfill blocks with index in {}..{}: {}",
+                                    start,
+                                    index,
+                                    fetch_err
+                                );
+                                Box::pin(once(async { Err(fetch_err) }).chain(item_stream))
+                            }
+                        }
+                    }
+                }
+                Err(upstream_error) => {
+                    // If we get an error, fetch one item.
+                    match fetcher.fetch_single(next_index) {
+                        Ok(future) => {
+                            prev_index = Some(next_index);
+                            Box::pin(future.into_stream())
+                        }
```

Comment:
I wonder how this would cascade, what is the cost of a single backfill? Might this get expensive if a node is having a longer term failure + what should we do in case of a longer term failure? I think at some point we'll have to do more fine grained error handling based upon exactly what error gets passed down the line.

Reply:
Yeah, I expect we'll evolve the error handling with usage.